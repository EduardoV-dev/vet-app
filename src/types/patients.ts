export interface IPatient {
    pet_name: string;
    owner_name: string;
    owner_email: string;
    discharge_date: string;
    symptoms: string;

    id?: string;
}
