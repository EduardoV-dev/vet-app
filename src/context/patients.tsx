import * as React from 'react';
import uniqid from 'uniqid';
import { IPatient } from '../types/patients';

/** Key used for setting and getting items from localStorage related to patients */
const LS_PATIENTS_KEY = 'vet-patients';

interface IPatientsContext {
    patients: IPatient[];
    selectedPatientId: string | null;

    deletePatient: (id: string) => void;
    editPatient: (patient: IPatient, id: string) => void;
    savePatient: (patient: IPatient) => void;
    selectPatientId: (id: string | null) => void;
}

export const PatientsContext = React.createContext({} as IPatientsContext);

/* --- Utils --- */

const saveInLocalStorage = (data: IPatient[]): void =>
    localStorage.setItem(LS_PATIENTS_KEY, JSON.stringify(data));

/* --- Provider Component --- */

const Provider = (props: { children: React.ReactNode }) => {
    /* --- Hooks --- */

    const [selectedPatientId, setSelectedPatientId] =
        React.useState<string | null>(null);
    const [patients, setPatients] = React.useState<IPatient[]>(() =>
        /** Gets patients from localStorage, returns [] if no patients are found. */

        JSON.parse(localStorage.getItem(LS_PATIENTS_KEY) || '[]'),
    );

    /* --- Handlers --- */

    /**
     * Saves a patient by passing its data, saves in localStorage and context state
     * @param patient - patient data
     * @returns {void}
     */
    const savePatient = (patient: IPatient): void =>
        setPatients((prevPatients) => {
            const newPatients: IPatient[] = [
                {
                    ...patient,
                    id: uniqid(),
                },
                ...prevPatients,
            ];

            saveInLocalStorage(newPatients);

            return newPatients;
        });

    /**
     * Deletes a patient register from localStorage and State by passing the id of the patient.
     * @param id - Patient id to be deleted.
     * @returns {void}
     */
    const deletePatient = (id: string): void => {
        if (!confirm('Do you want to delete the patient?')) return;

        setPatients((prevPatients) => {
            const newPatients: IPatient[] = prevPatients.filter(
                (patient) => patient.id !== id,
            );

            saveInLocalStorage(newPatients);

            return newPatients;
        });
    };

    /**
     * Select patient id for mapping the patient data in the form
     * @param id - the id of the patient to select for updating later. If the selected patient
     * id is the same as the one provided in the param will be unselected, else, will be selected.
     * @returns {void}
     */
    const selectPatientId = (id: string | null): void =>
        setSelectedPatientId((selectedId) => (selectedId === id ? null : id));

    /**
     * After selecting the patient id, use this function for editting the selected patient by passing
     * its new data, update happens in localStorage and state
     * @param patient - patient data to be editted
     * @returns
     */
    const editPatient = (patient: IPatient): void => {
        if (!selectedPatientId) return;

        setPatients((prevPatients) => {
            const newPatients: IPatient[] = prevPatients.map((pat) =>
                pat.id === selectedPatientId ? { ...patient } : pat,
            );

            saveInLocalStorage(newPatients);

            return newPatients;
        });
    };

    const value: IPatientsContext = {
        patients,
        selectedPatientId,

        deletePatient,
        editPatient,
        savePatient,
        selectPatientId,
    };

    return <PatientsContext.Provider {...{ value }} {...props} />;
};

export default Provider;
