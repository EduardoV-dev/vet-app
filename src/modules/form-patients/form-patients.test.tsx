import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { PatientsContext } from '../../context/patients';
import { IPatient } from '../../types/patients';
import FormPatients from '.';

describe('<FormPatients />', () => {
    const savePatient = jest.fn();
    const editPatient = jest.fn();

    let petInput: HTMLElement;
    let ownerNameInput: HTMLElement;
    let ownerEmailInput: HTMLElement;
    let dischargeDateInput: HTMLElement;
    let symptomsInput: HTMLElement;
    let submitButton: HTMLElement;

    const formData: IPatient = {
        pet_name: 'John',
        owner_name: 'Doe',
        owner_email: 'john@doe.com',
        discharge_date: '2022-07-05',
        symptoms: 'Lorem ipsum dolor sit amet',
    };

    /**
     * fills form inputs with the formData
     */
    const fillForm = (): void => {
        fireEvent.change(petInput, {
            target: { value: formData.pet_name },
        });
        fireEvent.change(ownerNameInput, {
            target: { value: formData.owner_name },
        });
        fireEvent.change(ownerEmailInput, {
            target: { value: formData.owner_email },
        });
        fireEvent.change(dischargeDateInput, {
            target: { value: formData.discharge_date },
        });
        fireEvent.change(symptomsInput, {
            target: { value: formData.symptoms },
        });
    };

    beforeEach(() => {
        render(
            <PatientsContext.Provider
                value={{
                    patients: [],
                    selectedPatientId: null,

                    savePatient,
                    selectPatientId: jest.fn(),
                    deletePatient: jest.fn(),
                    editPatient,
                }}
            >
                <FormPatients />
            </PatientsContext.Provider>,
        );

        petInput = screen.getByPlaceholderText(/enter pet's name/i);
        ownerNameInput = screen.getByPlaceholderText(/enter owner's name/i);
        ownerEmailInput = screen.getByPlaceholderText(/enter owner's email/i);
        dischargeDateInput = screen.getByTestId('dischargeDate');
        symptomsInput = screen.getByPlaceholderText(/describe pet's symptoms/i);
        submitButton = screen.getByRole('button', { name: /add patient/i });
    });

    it('Should render correctly with default input values', () => {
        expect((petInput as any).value).toBe('');
        expect((ownerNameInput as any).value).toBe('');
        expect((ownerEmailInput as any).value).toBe('');
        expect((dischargeDateInput as any).value).toBe('');
        expect((symptomsInput as any).value).toBe('');
        expect(submitButton).not.toHaveStyle('opacity: 0.8');
    });

    it('Should fill the form', async () => {
        fillForm();

        expect((petInput as any).value).toBe(formData.pet_name);
        expect((ownerNameInput as any).value).toBe(formData.owner_name);
        expect((ownerEmailInput as any).value).toBe(formData.owner_email);
        expect((dischargeDateInput as any).value).toBe(formData.discharge_date);
        expect((symptomsInput as any).value).toBe(formData.symptoms);
    });

    it('Should not submit the form if all the fields are not satisfied', () => {
        expect((petInput as any).value).toBe('');

        fireEvent.click(submitButton);

        expect(savePatient).not.toHaveBeenCalled();
    });

    it('Should submit the form', async () => {
        fillForm();

        fireEvent.click(submitButton);

        await waitFor(() => expect(savePatient).toHaveBeenCalledWith(formData));
    });
});
