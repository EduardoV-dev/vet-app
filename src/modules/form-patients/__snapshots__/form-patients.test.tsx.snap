// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormPatients /> Should render correctly with default input values 1`] = `
<div>
  <section
    class="flex flex-col items-center"
    style="opacity: 0; transform: translate3d(0,50px,0) scale(0);"
  >
    <h2>
      Patients
    </h2>
    <h3>
      Add patients &
       
      <span
        class="highlight"
      >
        Manage them
      </span>
    </h3>
    <section
      class="py-8 px-4 shadow-md w-9/12 mx-auto bg-white mt-10 "
    >
      <form
        action="#"
      >
        <div
          class="flex flex-col mb-4"
        >
          <label
            class="uppercase font-bold"
            for="petName"
          >
            Pet Name
          </label>
          <input
            class="rounded p-2 border-2 focus:border-blue-100 mt-1"
            id="petName"
            name="pet_name"
            placeholder="Enter pet's name"
            value=""
          />
        </div>
        <div
          class="flex flex-col mb-4"
        >
          <label
            class="uppercase font-bold"
            for="ownerName"
          >
            Owner's Name
          </label>
          <input
            class="rounded p-2 border-2 focus:border-blue-100 mt-1"
            id="ownerName"
            name="owner_name"
            placeholder="Enter owner's name"
            value=""
          />
        </div>
        <div
          class="flex flex-col mb-4"
        >
          <label
            class="uppercase font-bold"
            for="ownerEmail"
          >
            Owner's Email
          </label>
          <input
            class="rounded p-2 border-2 focus:border-blue-100 mt-1"
            id="ownerEmail"
            name="owner_email"
            placeholder="Enter owner's email"
            type="email"
            value=""
          />
        </div>
        <div
          class="flex flex-col mb-4"
        >
          <label
            class="uppercase font-bold"
            for="dischargeDate"
          >
            Discharge Date
          </label>
          <input
            class="rounded p-2 border-2 focus:border-blue-100 mt-1"
            data-testid="dischargeDate"
            id="dischargeDate"
            name="discharge_date"
            type="date"
            value=""
          />
        </div>
        <div
          class="flex flex-col mb-4"
        >
          <label
            class="uppercase font-bold"
            for="symptoms"
          >
            Symptoms
          </label>
          <textarea
            class="rounded p-2 border-2 focus:border-blue-100 mt-1 min-h-[100px] max-h-[200px]"
            id="symptoms"
            name="symptoms"
            placeholder="Describe pet's symptoms"
            type="textarea"
          />
        </div>
        <button
          class="font-bold uppercase py-2 px-10 mt-2 bg-blue-700 hover:bg-blue-600 text-white transition active:scale-95 w-full"
          type="submit"
        >
          Add Patient
        </button>
      </form>
    </section>
  </section>
</div>
`;
